clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchDocumentMapper.scala

```diff
@@ -1,29 +1,50 @@
 package org.broadinstitute.clio.server.dataaccess.elasticsearch
 
+import io.circe.{Encoder, Json}
+import org.broadinstitute.clio.util.generic.CaseClassMapper
+import org.broadinstitute.clio.util.model.{EntityId, UpsertId}
+
+import scala.reflect.ClassTag
+
 /**
   * Maps metadata to an Elasticsearch document.
   *
-  * @tparam ModelKey      The type of key.
-  * @tparam ModelMetadata The type of the metadata.
-  * @tparam Document      The type of the document.
+  * @tparam Key      The type of the TransferKey.
+  * @tparam Metadata The type of the TransferMetadata.
   */
-abstract class ElasticsearchDocumentMapper[ModelKey, ModelMetadata, Document] {
+class ElasticsearchDocumentMapper[Key <: Product: ClassTag, Metadata: ClassTag] (genId: () => UpsertId)(
+  implicit
+  private[elasticsearch] val keyEncoder: Encoder[Key],
+  private[elasticsearch] val metadataEncoder: Encoder[Metadata]
+) {
+  private val keyMapper = new CaseClassMapper[Key]
+  private val metadataMapper = new CaseClassMapper[Metadata]
 
   /**
-    * Returns an empty document for a key.
     *
-    * @param key The key.
-    * @return The newly initialized document.
+    * @param key      The TransferKey from which to create the document.
+    * @param metadata The TransferMetadata from which to create the document.
+    * @return
     */
-  def empty(key: ModelKey): Document
+  def document(key: Key, metadata: Metadata): Json = {
+    val keyVals = keyMapper.vals(key)
+    val metadataVals = metadataMapper.vals(metadata)
+    val bookkeeping = Map(
+      UpsertId.UpsertIdFieldName -> genId(),
+      EntityId.EntityIdFieldName -> Symbol(
+        key.productIterator.mkString(".")
+      )
+    )
```

Comment:
These steps aren't needed, we can go straight to JSON: `key.asJson`.

`Json.fromFields` should help with building up the bookkeeping object.